File: apps/api/middleware/tenant.go
package middleware

import (
	"fmt"
	"net/http"
	"strings"

	"github.com/gin-gonic/gin"
)

const TenantKey = "nest_tenant"
const ClaimsKey = "nest_claims"

// Claims represents the JWT claims subset Nest cares about
type Claims struct {
	Sub    string   `json:"sub"`
	Tenant string   `json:"tenant"`
	Scopes []string `json:"scopes"`
	Tier   string   `json:"tier"`
}

// TenantMiddleware enforces presence of tenant claim on every request.
// In P1 this validates a simple Bearer token; full OIDC wired in P2+.
func TenantMiddleware() gin.HandlerFunc {
	return func(c *gin.Context) {
		auth := c.GetHeader("Authorization")
		if !strings.HasPrefix(auth, "Bearer ") {
			c.AbortWithStatusJSON(http.StatusUnauthorized, nestError(
				"nest.auth.missing_token",
				"Authorization header with Bearer token required",
				c.Request.Header.Get("X-Request-ID"),
			))
			return
		}

		// TODO(P2): validate JWT signature via OIDC JWKS
		claims, err := parseToken(strings.TrimPrefix(auth, "Bearer "))
		if err != nil || claims.Tenant == "" {
			c.AbortWithStatusJSON(http.StatusForbidden, nestError(
				"nest.auth.missing_tenant",
				"JWT missing mandatory tenant claim",
				c.Request.Header.Get("X-Request-ID"),
			))
			return
		}

		c.Set(TenantKey, claims.Tenant)
		c.Set(ClaimsKey, claims)
		c.Next()
	}
}

// GetTenant extracts the validated tenant from context
func GetTenant(c *gin.Context) string {
	t, _ := c.Get(TenantKey)
	s, _ := t.(string)
	return s
}

// GetClaims extracts the validated claims from context
func GetClaims(c *gin.Context) *Claims {
	v, _ := c.Get(ClaimsKey)
	cl, _ := v.(*Claims)
	return cl
}

// RequireScope aborts if the token lacks the given scope
func RequireScope(scope string) gin.HandlerFunc {
	return func(c *gin.Context) {
		claims := GetClaims(c)
		if claims == nil {
			c.AbortWithStatusJSON(http.StatusForbidden, nestError(
				"nest.auth.scope_denied",
				"Authentication required",
				c.Request.Header.Get("X-Request-ID"),
			))
			return
		}
		for _, s := range claims.Scopes {
			if s == scope || s == "nest:*:admin" {
				c.Next()
				return
			}
		}
		c.AbortWithStatusJSON(http.StatusForbidden, nestError(
			"nest.auth.scope_denied",
			"Insufficient scope: "+scope+" required",
			c.Request.Header.Get("X-Request-ID"),
		))
	}
}

// parseToken is a stub for P1. In P2+ this validates via OIDC JWKS.
func parseToken(token string) (*Claims, error) {
	// P1: accept any token with format "tenant:scopes" for local dev
	// This is replaced by real OIDC validation in P2
	if token == "" {
		return nil, fmt.Errorf("empty token")
	}
	parts := strings.SplitN(token, ":", 3)
	claims := &Claims{
		Sub:    parts[0],
		Tenant: "",
		Scopes: []string{"nest:*:admin"},
		Tier:   "free",
	}
	if len(parts) >= 2 {
		claims.Tenant = parts[1]
	}
	if len(parts) >= 3 {
		claims.Tier = parts[2]
	}
	return claims, nil
}

func nestError(code, message, requestID string) gin.H {
	return gin.H{
		"code":      code,
		"message":   message,
		"requestId": requestID,
		"docsUrl":   "https://docs.nest.penguintech.io/errors/" + code,
	}
}

File: apps/api/middleware/tenant.py
"""Tenant middleware for authentication."""
import uuid
from dataclasses import dataclass
from functools import wraps
from typing import Optional

from quart import g, request
from werkzeug.exceptions import Unauthorized, Forbidden


@dataclass(slots=True)
class Claims:
    """JWT claims subset Nest cares about."""

    sub: str
    tenant: str
    scopes: list[str]
    tier: str


TENANT_KEY = "nest_tenant"
CLAIMS_KEY = "nest_claims"


def nest_error(code: str, message: str, request_id: str) -> dict:
    """Generate a Nest API error response."""
    return {
        "code": code,
        "message": message,
        "requestId": request_id,
        "docsUrl": f"https://docs.nest.penguintech.io/errors/{code}",
    }


def parse_token(token: str) -> Optional[Claims]:
    """Parse a simple Bearer token format: sub:tenant:tier.

    P1 stub; P2+ will use OIDC JWKS validation.
    """
    if not token:
        return None
    parts = token.split(":", 3)
    claims = Claims(
        sub=parts[0] if len(parts) > 0 else "",
        tenant=parts[1] if len(parts) > 1 else "",
        scopes=["nest:*:admin"],
        tier=parts[2] if len(parts) > 2 else "free",
    )
    return claims


async def tenant_middleware():
    """Middleware that enforces presence of tenant claim.

    Runs before all /api/v1/tenants/ routes.
    Stores claims in g context for per-request access.
    """
    auth = request.headers.get("Authorization", "")
    request_id = request.headers.get("X-Request-ID", str(uuid.uuid4()))

    if not auth.startswith("Bearer "):
        raise Unauthorized(
            response=nest_error(
                "nest.auth.missing_token",
                "Authorization header with Bearer token required",
                request_id,
            )
        )

    token = auth[7:]  # Remove "Bearer " prefix
    claims = parse_token(token)

    if not claims or not claims.tenant:
        raise Forbidden(
            response=nest_error(
                "nest.auth.missing_tenant",
                "JWT missing mandatory tenant claim",
                request_id,
            )
        )

    # Store in g for request scope
    g.tenant = claims.tenant
    g.claims = claims
    g.request_id = request_id


def get_tenant() -> str:
    """Extract the validated tenant from request context."""
    return getattr(g, "tenant", "")


def get_claims() -> Optional[Claims]:
    """Extract the validated claims from request context."""
    return getattr(g, "claims", None)


def require_scope(scope: str):
    """Decorator to require a specific scope on the request."""

    def decorator(f):
        @wraps(f)
        async def decorated_function(*args, **kwargs):
            claims = get_claims()
            request_id = getattr(g, "request_id", str(uuid.uuid4()))

            if not claims:
                raise Forbidden(
                    response=nest_error(
                        "nest.auth.scope_denied",
                        "Authentication required",
                        request_id,
                    )
                )

            # Check if token has the scope or admin scope
            has_scope = False
            for s in claims.scopes:
                if s == scope or s == "nest:*:admin":
                    has_scope = True
                    break

            if not has_scope:
                raise Forbidden(
                    response=nest_error(
                        "nest.auth.scope_denied",
                        f"Insufficient scope: {scope} required",
                        request_id,
                    )
                )

            return await f(*args, **kwargs)

        return decorated_function

    return decorator
